mava/components/executing/action_selection.py

```diff
@@ -112,31 +106,119 @@ def on_execution_select_actions(self, executor: SystemExecutor) -> None:
 
-        # executor.store.observation set by Executor
-        observation = utils.add_batch_dim(executor.store.observation.observation)
-        rng_key, executor.store.base_key = jax.random.split(executor.store.base_key)
+            Args:
+                observations : all obs.
+                current_params : current params.
+                key : prng_key.
 
-        # TODO (dries): We are currently using jit in the networks per agent.
-        # We can also try jit over all the agents in a for loop. This would
-        # allow the jit function to save us even more time.
-        executor.store.action_info, executor.store.policy_info = network.get_action(
-            observation,
-            rng_key,
-            utils.add_batch_dim(executor.store.observation.legal_actions),
+            Returns:
+                action info, policy info and new prng key.
+            """
+            actions_info, policies_info = {}, {}
+            # TODO Look at tree mapping this forloop.
+            # Since this is jitted, compiling a forloop with lots of agents could take
+            # long, we should vectorize this.
+            for agent, observation in observations.items():
+                network = networks[agent_net_keys[agent]]
+                actions_info[agent], policies_info[agent], key = select_action(
+                    observation=observation,
+                    current_params=current_params[agent_net_keys[agent]],
+                    network=network,
+                    key=key,
+                )
+            return actions_info, policies_info, key
+
+        executor.store.select_actions_fn = jax.jit(select_actions)
+
+
+class RecurrentExecutorSelectAction(ExecutorSelectAction):
+    def __init__(
+        self,
+        config: SimpleNamespace = SimpleNamespace(),
+    ):
+        """Component defines hooks for the executor selecting actions.
+
+        Args:
+            config: SimpleNamespace.
+        """
+        self.config = config
+
+    # Select actions
+    def on_execution_select_actions(self, executor: SystemExecutor) -> None:
+        """Select actions for each agent and save info in store.
+
+        Args:
+            executor: SystemExecutor.
+
+        Returns:
+            None.
+        """
+
+        # Dict with params per network
+        current_agent_params = {
+            network: executor.store.networks[network].get_params()
+            for network in executor.store.agent_net_keys.values()
+        }
+
+        (
+            executor.store.actions_info,
+            executor.store.policies_info,
+            executor.store.policy_states,
+            executor.store.base_key,
+        ) = executor.store.select_actions_fn(
+            executor.store.observations,
+            current_agent_params,
+            executor.store.policy_states,
+            executor.store.base_key,
         )
 
     def on_execution_init_end(self, executor: SystemExecutor) -> None:
```

Comment:
We can maybe have a chat about this. @KaleabTessera what do you think?

Reply:
Happy to have an offline discussion on this but this would in any case be something for another PR. So going to approve this so long. Thanks @DriesSmit 😉

Reply:
Thanks :slightly_frowning_face: I will create an issue for this :+1: